src/style-spec/validate/validate_expression.js

```diff
@@ -14,9 +14,9 @@ export default function validateExpression(options: any) {
         });
     }
 
-    if (options.expressionContext === 'property' && options.propertyKey === 'text-font' &&
+    if (options.expressionContext === 'property' && (options.propertyKey === 'text-font' || options.propertyKey === 'line-pattern') &&
         (expression.value: any)._styleExpression.expression.possibleOutputs().indexOf(undefined) !== -1) {
-        return [new ValidationError(options.key, options.value, 'Invalid data expression for "text-font". Output values must be contained as literals within the expression.')];
+        return [new ValidationError(options.key, options.value, `Invalid data expression for "${options.propertyKey}". Output values must be contained as literals within the expression.`)];
```

Comment:
Is it true that this in fact applies to any cross-faded-data-driven property?  If so, let's do this check using use the `property-type` value instead of checking for each -pattern property... though unfortunately there'll be a bit of a naming clash adding this to the validation options: https://github.com/mapbox/mapbox-gl-js/blob/5b3af8e1515ee311967661b5aabdab65f1376e36/src/style-spec/validate/validate_property.js#L62